natlib/svcd.c

```diff
@@ -30,6 +31,54 @@ static const LUA_REG_TYPE svcd_meta_map[] =
 // Maintainer: Michael Andersen <michael@steelcode.com>
 /////////////////////////////////////////////////////////////
 
+//Lua: storm.n.svcd_advert_received(pay, srcip, srcport)
+//Dispatch an advertisment
+
+int svcd_advert_received(lua_State *L) 
+{
+	char *pay = luaL_checkstring(L, 1);
+	char *srcip = luaL_checkstring(L, 2);
+	int srcport = luaL_checknumber(L, 3);
+	lua_pushlightfunction(L, libmsgpack_mp_unpack);      
+        lua_pushstring(L,pay); 
+	lua_call(L,1,1);	
+        printf("Service advertisment %s\n", srcip);
+    	lua_pushvalue(L, -1);
+    	lua_pushnil(L);
+    	while (lua_next(L, -2))
+    	{
+        	lua_pushvalue(L, -2);
+        	const char *k = lua_tostring(L, -1);
+        	if (strcmp(k, "id")==0)
+		{	
+        		const char *v = lua_tostring(L, -2);
+	        	printf("ID = %s\n", v);	
+		}
+      		else
+		{
+			unsigned int k = lua_tonumber(L,-1);
+        		printf("0x%04x:\n",k);
+ 
+	    		lua_pushvalue(L, -2);
+    			lua_pushnil(L);
+    			while (lua_next(L, -2))
+    			{
+        			lua_pushvalue(L, -2);
+        			int kk = lua_tonumber(L, -1);
+        			unsigned int vv = lua_tonumber(L, -2);
+				printf("   >%d: 0x%04x\n", kk, vv);	        		      
+				lua_pop(L, 2);
```

Comment:
We are popping the key and the value, so we can parse the next key-value pair from the table. Not sure what your question is.